zilliqa/src/blockhooks.rs

```diff
@@ -0,0 +1,118 @@
+use crate::message::IntershardCall;
+use crate::state::contract_addr;
+use crate::{contracts, state::Address, transaction::TransactionReceipt};
+use anyhow::Result;
+use ethabi::{Event, Log, RawLog, Token};
+use tracing::warn;
+
+#[derive(Debug, Default)]
+pub struct BlockHooks();
+
+impl BlockHooks {
+    fn filter_receipts(
+        &self,
+        receipts: Vec<TransactionReceipt>,
+        event: Event,
+        emitter: Address,
+    ) -> Result<Vec<Log>> {
+        let logs: Result<Vec<_>, _> = receipts
+            .into_iter()
+            .flat_map(|receipt| receipt.logs)
+            .filter(|log| log.address == emitter && log.topics[0] == event.signature())
+            .map(|log| {
+                event
+                    // parse_log_whole can't be used here because it doesn't seem to work
```

Comment:
If the right events are used, there should be no danger, but `parse_log_whole` would add a layer of assurance that we're definitely parsing exactly the right event. (Or, to be precise, at least an event which is exactly the length we expect, though you can't have 100% type safety when parsing a binary blob.)

>Do you have any more info on the error occurring with `parse_log_whole`?
Basically, `ethabi` parses events using a cursor that's incremented after each field, and after going through every field in the event `parse_log_whole` throws an error if `cursor_position != data_length`.
However, dynamic length fields - like `bytes` in the intershard call event - are encoded in the solidity ABI as just a pointer to the actual location of the data, which is in fact is appended as a length-prefixed blob _after_ all the fixed-length fields. Kind of like a "heap" area located after the known-size "stack" of fields. And `ethabi`'s cursor position just goes through the stack, so at the end of the parsing it's _not_ located at the end of the data because you've got the heap afterwards, and it throws an error without taking into account the fact that it already parsed the heap data as part of parsing the `bytes` field earlier.

Reply:
I see, thanks for the explanation. This seems like a reasonable solution for now in that case.